### PR TITLE
Fix the gradient norm reported by the trainer

### LINKED ISSUE #518: Incorrect gradient norm reported in Trainer (for epochs > 1)
The reported gradient norm incorrectly lags due to a missing metric reset.

**Fix**

The variable tracking the gradient norm is updated every iteration (multiple times per epoch) but it is never reset:

https://github.com/deepinv/deepinv/blob/64703372b42d699248ec7acfb576c5626880f1db/deepinv/training/trainer.py#L420-L429

One way to fix the bug is to add this code snippet:

```python
self.check_grad_val.reset()
```

to the method `reset_metrics`:

https://github.com/deepinv/deepinv/blob/64703372b42d699248ec7acfb576c5626880f1db/deepinv/training/trainer.py#L921-L940

**Reproducing**

```python
# We run the trainer while controlling the gradient norm to show that the
# trainer reports the wrong value. We do so by mocking torch.Tensor.backward
# and making it set gradients with an overall norm of 1.0 for the 1st epoch,
# 2.0 for the 2nd, and so on.

import deepinv as dinv
from torch.utils.data import DataLoader
import torch
from pathlib import Path
from torchvision import transforms
from deepinv.utils.demo import load_dataset

from unittest.mock import patch
import math


physics = dinv.physics.Inpainting(tensor_size=(3, 16, 16), mask=0.5, device="cpu")

train_dataset = dinv.datasets.CBSD68(
    "data/Set14",
    transform=transforms.Compose(
        [
            transforms.ToTensor(),
            transforms.Resize((16, 16)),
        ]
    ),
    download=True,
)
train_dataloader = DataLoader(train_dataset, batch_size=1, shuffle=False)

backbone = dinv.models.UNet(in_channels=3, out_channels=3, scales=2)
model = dinv.models.ArtifactRemoval(backbone)

losses = [
    dinv.loss.SupLoss(),
]

optimizer = torch.optim.AdamW(model.parameters(), lr=1e-3)

trainer = dinv.Trainer(
    model,
    device="cpu",
    save_path="ckpts",
    verbose=True,
    show_progress_bar=False,
    physics=physics,
    epochs=10,
    losses=losses,
    optimizer=optimizer,
    train_dataloader=train_dataloader,
    online_measurements=True,
    check_grad=True,
)

call_count = 0
calls_per_epoch = math.ceil(len(train_dataset) / train_dataloader.batch_size)


def mock_fn(self):
    global call_count
    epoch = 1 + call_count // calls_per_epoch
    call_count += 1

    # 1. Fill in the gradients with random values.
    for p in model.parameters():
        if p.requires_grad:
            p.grad = torch.ones_like(p)

        # 2. Compute the norm of the gradients.
        # from https://discuss.pytorch.org/t/check-the-norm-of-gradients/27961/7
        grads = [
            p.grad.detach().flatten() for p in model.parameters() if p.grad is not None
        ]
        grads = torch.cat(grads)
        norm = grads.norm()
        norm = norm.item()

    # 3. Rescale the gradients so that the gradient norm is equal to 1.0 for
    # the 1st epoch, to 2.0 for the 2nd, and so on.
    for p in model.parameters():
        if p.requires_grad:
            p.grad = epoch * p.grad / norm


with patch.object(torch.Tensor, "backward", mock_fn):
    trainer.train()
```

**Observed results**

```
The model has 446019 trainable parameters
Train epoch 0: TotalLoss=1.929, gradient_norm=1.0, PSNR=-1.651
Train epoch 1: TotalLoss=8.109, gradient_norm=1.5, PSNR=-8.9
Train epoch 2: TotalLoss=10.633, gradient_norm=2.0, PSNR=-10.14
Train epoch 3: TotalLoss=5.071, gradient_norm=2.5, PSNR=-6.078
Train epoch 4: TotalLoss=2.568, gradient_norm=3.0, PSNR=-0.068
Train epoch 5: TotalLoss=0.584, gradient_norm=3.5, PSNR=2.395
Train epoch 6: TotalLoss=0.745, gradient_norm=4.0, PSNR=1.321
Train epoch 7: TotalLoss=0.929, gradient_norm=4.5, PSNR=0.355
Train epoch 8: TotalLoss=1.134, gradient_norm=5.0, PSNR=-0.52
Train epoch 9: TotalLoss=1.361, gradient_norm=5.5, PSNR=-1.317
```

**Expected results**

The gradient norm should be 1.0 for the 1st epoch, 2.0 for the 2nd epoch, and so on. Indeed, it is what we get when applying the fix:

```
The model has 446019 trainable parameters
Train epoch 0: TotalLoss=2.696, gradient_norm=1.0, PSNR=-3.434
Train epoch 1: TotalLoss=9.387, gradient_norm=2.0, PSNR=-9.573
Train epoch 2: TotalLoss=11.426, gradient_norm=3.001, PSNR=-10.438
Train epoch 3: TotalLoss=5.3, gradient_norm=4.0, PSNR=-6.224
Train epoch 4: TotalLoss=2.481, gradient_norm=4.999, PSNR=0.261
Train epoch 5: TotalLoss=0.531, gradient_norm=6.001, PSNR=2.816
Train epoch 6: TotalLoss=0.685, gradient_norm=6.999, PSNR=1.692
Train epoch 7: TotalLoss=0.861, gradient_norm=8.001, PSNR=0.685
Train epoch 8: TotalLoss=1.06, gradient_norm=8.999, PSNR=-0.223
Train epoch 9: TotalLoss=1.279, gradient_norm=9.998, PSNR=-1.047
``

### PR DESCRIPTION
**Pull request by jscanvic**
Fixes #518 

### Checks to be done before submitting your PR
- [x] `python3 -m pytest tests/` runs successfully.
- [x] `black .` runs successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

**Review by tachella — APPROVED** [review]
looks good to me, thanks for the fix and the test checking errors
